src/editors/containers/ProblemEditor/data/OLXParser.js

```diff
@@ -91,18 +92,74 @@ export class OLXParser {
       },
       processEntities: false,
     };
+    const richTextBuilderOptions = {
+      ignoreAttributes: false,
+      numberParseOptions: {
+        leadingZeros: false,
+        hex: false,
+      },
+      preserveOrder: true,
+      processEntities: false,
+    };
     // There are two versions of the parsed XLM because the question requires the order of the
     // parsed data to be preserved. However, all the other widgets need the data grouped by
     // the wrapping tag.
-    const questionParser = new XMLParser(questionOptions);
+    const richTextParser = new XMLParser(richTextOptions);
     const parser = new XMLParser(parserOptions);
     this.builder = new XMLBuilder(builderOptions);
+    this.richTextBuilder = new XMLBuilder(richTextBuilderOptions);
     this.parsedOLX = parser.parse(olxString);
-    this.parsedQuestionOLX = questionParser.parse(olxString);
+    this.richTextOLX = richTextParser.parse(olxString);
     if (_.has(this.parsedOLX, 'problem')) {
       this.problem = this.parsedOLX.problem;
-      this.questionData = this.parsedQuestionOLX[0].problem;
+      this.questionData = this.richTextOLX[0].problem;
+      this.richTextProblem = this.richTextOLX[0].problem;
+    }
+  }
+
+  /** getPreservedAnswersAndFeedback(problemType, widgetName, option)
+   * getPreservedAnswersAndFeedback takes a problemType, widgetName, and a valid option. The
+   * olx for the given problem type and widget is parsed. Depending on the
+   * @param {string} problemType - string of the olx problem type
+   * @param {string} widgetName - string of the wrapping tag name
+   *                              (optioninput, choicegroup, checkboxgroup, addition_answer)
+   * @param {string} option - string of the type of answers (choice, option)
+   * @return {array} array containing answer objects and possibly an array of grouped feedback
+   */
+  getPreservedAnswersAndFeedback(problemType, widgetName, option) {
+    const preservedProblem = this.richTextProblem;
+    const isChoiceProblem = problemType !== ProblemTypeKeys.NUMERIC ? problemType !== ProblemTypeKeys.TEXTINPUT : false;
+    const preservedAnswers = [];
+    let correctFeedbackTag = option;
+    let incorrectFeedbackTag;
+    if (problemType === ProblemTypeKeys.TEXTINPUT) {
+      [correctFeedbackTag, incorrectFeedbackTag] = option;
     }
+    preservedProblem.forEach(wrapperTag => {
+      const wrapperTagKeys = Object.keys(wrapperTag);
+      if (wrapperTagKeys.includes(problemType)) {
```

Comment:
This function has a very high complexity. There are multiple nested if statements and for loops and it is not possible for me to follow along with the code without spending a large amount of time analyzing it. It seems to me to make this more readable and maintainable, it is helpful to split this up into several well-named functions and variables, and avoid nested ifs / for loops.

Reply:
I will mark this as optional. You can decide whether you want to do this here or not. I would of course welcome any improvements of readability, but the parser is already very complex anyway and we can also leave refactoring for another time. With the great amount of work you did in this issue I'm sure you'll be glad to get this merged.